pinot-controller/src/test/java/org/apache/pinot/controller/api/upload/ZKOperatorTest.java

```diff
@@ -64,11 +64,34 @@ public void testCompleteSegmentOperations()
     when(segmentMetadata.getCrc()).thenReturn("12345");
     when(segmentMetadata.getIndexCreationTime()).thenReturn(123L);
     HttpHeaders httpHeaders = mock(HttpHeaders.class);
+
+    // Test if Zk segment metadata is removed if exception is thrown when moving segment to final location.
+    try {
+      // Create mock finalSegmentLocationURI and currentSegmentLocation.
+      URI finalSegmentLocationURI =
+          URIUtils.getUri("mockPath", OFFLINE_TABLE_NAME, URIUtils.encode(segmentMetadata.getName()));
+      File currentSegmentLocation = new File(new File("foo/bar"), "mockChild");
+
+      zkOperator
+          .completeSegmentOperations(OFFLINE_TABLE_NAME, segmentMetadata, finalSegmentLocationURI,
+              currentSegmentLocation, true, httpHeaders, "downloadUrl",
+              true, "crypter");
+      fail();
+    } catch (Exception e) {
+      // Expected
+    }
+
+    // Add a tiny sleep to give some time for the segment Zk entry to be deleted.
+    Thread.sleep(5000L);
```

Comment:
We can use waitForCondition method to validate instead of sleeping a fixed amount of time.

Reply:
Done. Thanks!